### PR TITLE
rw spin lock

### PR DESCRIPTION
**Pull request by havardpe**
still only experimental; both the lock itself and its benchmarking
spin-off: Nexus utility for multi-threaded testing and benchmarking

@vekterli please review
@geirst FYI (Nexus spin-off)

### CODE REVIEW

**Comment by vekterli** [comment]
PR build failed with
```
15:58:44 /sd/workspace/src/github.com/vespa-engine/vespa/vespalib/src/tests/nexus/nexus_test.cpp: In lambda function:
15:58:44 /sd/workspace/src/github.com/vespa-engine/vespa/vespalib/src/tests/nexus/nexus_test.cpp:70:9: error: ‘EXPECT_EQ’ was not declared in this scope
15:58:44    70 |         EXPECT_EQ(ctx.num_threads(2));
15:58:44       |         ^~~~~~~~~
```

**Review by vekterli — DISMISSED** [review]
:+1: Side note: I didn't stare too hard at the spinlock logic itself since it looked unchanged since the previous PR (where I did stare hard).

File: vespalib/src/tests/nexus/nexus_test.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+// Copyright Yahoo. Licensd under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <vespa/vespalib/test/nexus.h>
+#include <vespa/vespalib/gtest/gtest.h>
+#include <vespa/vespalib/util/require.h>
+
+using namespace vespalib::test;
+
+TEST(NexusTest, run_void_tasks) {
+    std::atomic<size_t> value;
```
[the remark this answers is not in the document]
Comment (havardpe) [reply vespalib/src/tests/nexus/nexus_test.cpp:10]: done

File: vespalib/src/tests/nexus/nexus_test.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    auto task = [&value](Nexus &) {
+                    return value.fetch_add(1, std::memory_order_relaxed) + 1;
+                };
+    Nexus ctx(10);
+    EXPECT_EQ(ctx.run(task, Nexus::merge_sum()), 55);
+    EXPECT_EQ(value, 10);
+}
+
+TEST(NexusTest, run_inline_voted_loop) {
```
[the remark this answers is not in the document]
Comment (havardpe) [reply vespalib/src/tests/nexus/nexus_test.cpp:42]: done

**Review by vekterli — APPROVED** [review]

**Pull request merged**